controllers/nstemplatetierrevisioncleanup/nstemplatetier_revision_cleanup_controller.go

```diff
@@ -0,0 +1,144 @@
+	Client runtimeclient.Client
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	// fetch the NSTemplateTier tier
+	ttr := &toolchainv1alpha1.TierTemplateRevision{}
+	if err := r.Client.Get(ctx, request.NamespacedName, ttr); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("TierTemplateRevision not found")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("unable to get the current TierTemplateRevision: %w", err)
+	}
+
+	//there is no point in fetching  the NStemplateTier, if the TTR  is just created
+	//get the tier template revision creation time stamp and the duration
+	timeSinceCreation := time.Since(ttr.GetCreationTimestamp().Time)
+
+	//the ttr age should be greater than 30 seconds
+	if timeSinceCreation < DeletionTimeThreshold {
+		requeAfter := DeletionTimeThreshold - timeSinceCreation
+		return reconcile.Result{RequeueAfter: requeAfter}, nil
+	}
+
+	//check if there is tier-name label available
+	tierName, ok := ttr.GetLabels()[toolchainv1alpha1.TierLabelKey]
+	if !ok {
+		//if tier-name label is not found we should delete the TTR
+		if err := r.deleteTTR(ctx, ttr); err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+	// fetch the related NSTemplateTier tier
+	tier := &toolchainv1alpha1.NSTemplateTier{}
+	if err := r.Client.Get(ctx, types.NamespacedName{
+		Namespace: ttr.GetNamespace(),
+		Name:      tierName,
+	}, tier); err != nil {
+		if errors.IsNotFound(err) {
+			// if there is no related nstemplate tier, we can delete the TTR directly
+			if err := r.deleteTTR(ctx, ttr); err != nil {
+				return reconcile.Result{}, err
+			}
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("unable to get the current NSTemplateTier: %w", err)
+	}
+
+	// verify that the tier template revision which is unused(not referenced by any tiers
+	// and all the spaces are up-to-date)
+	isTTrUnused, err := r.verifyUnusedTTR(ctx, tier, ttr)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Delete the unused revision
+	if isTTrUnused {
+		return reconcile.Result{}, r.deleteTTR(ctx, ttr)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// verifyUnusedTTR function verifies that the tier template revision
+// is not present/referenced in status.revisions field, and there are no outdated spaces.
+func (r *Reconciler) verifyUnusedTTR(ctx context.Context, nsTmplTier *toolchainv1alpha1.NSTemplateTier,
+	rev *toolchainv1alpha1.TierTemplateRevision) (bool, error) {
+
+	logger := log.FromContext(ctx)
+	//check if the ttr name is present status.revisions
+	for _, ttStatusRev := range nsTmplTier.Status.Revisions {
+		if ttStatusRev == rev.Name {
+			logger.Info("the revision is still being referenced in status.revisions", "ttr.name", rev.Name)
```

Comment:
two things:
* improve the message
* there is no need to reference the name of TTR - this is already included in the logger from the context
```suggestion
			logger.Info("the TierTemplateRevision is still being referenced in NSTemplateTier.Status.Revisions")
```

Reply:
i added the ttr name as a part of this comment , https://github.com/codeready-toolchain/host-operator/pull/1153#discussion_r1987731526 , changed it now

Reply:
my bad, I forgot that it's in the context 🤦‍♂️ . Sorry for the confusion.

Reply:
i blindly followed the suggestion rather than questioning /understanding it. 😕